### PR TITLE
Refactored FAR and Rename TreeView UI

### PR DESCRIPTION
**Pull request by Colengms**
Switch to a common TreeView object for all tree node types, across FAR and Rename.

Removed renameModel and renameDataProvider.  referencesModel and referencesDataProvider now handle rename as well.

Added group-by to rename candidate references.  Remembers user's last group-by setting for subsequent FAR or Rename.  (not persistent across launches)

Fixed a bug if cancel occurred before the request is sent, which would stall FAR/Rename requests.

Replace rejects with empty resolves to avoid error messages from VS Code.

### CODE REVIEW

File: Extension/src/LanguageServer/referencesView.ts [hunk 1]
```diff
@@ -57,16 +65,15 @@ export class FindAllRefsView {
             }
         }
 
-        // Get files with pending references items (location of reference is pending)
-        let fileReferences: FileItem[] = this.referenceViewProvider.getFilesWithPendingReferences();
+        // // Get files with pending references items (location of reference is pending)
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/referencesView.ts]: Two // ?

**Review by sean-mcmanus — APPROVED** [review]

**Comment by michelleangela** [comment]
Thanks for refactoring. Could we hold off on merging this until I get back? I would like a chance to review it and add input.

File: Extension/src/LanguageServer/referencesProvider.ts [hunk 2]
```diff
@@ -47,119 +47,118 @@ function getReferenceItemIconPath(type: ReferenceType, isCanceled: boolean): { l
...
-    private readonly _onDidChangeTreeData = new vscode.EventEmitter<TreeObject>();
+export class ReferenceDataProvider implements vscode.TreeDataProvider<TreeNode> {
+    private referencesModel: ReferencesModel;
+    private readonly _onDidChangeTreeData = new vscode.EventEmitter<TreeNode>();
     readonly onDidChangeTreeData = this._onDidChangeTreeData.event;
-    private groupByFile: boolean = true;
 
-    constructor() {
-        vscode.commands.executeCommand('setContext', 'refView.isGroupedByFile', true);
+    constructor(readonly isCandidates: boolean) {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesProvider.ts]: >isCandidates [](start = 25, length = 12)

nit: how about "isRenameCandidates" so it's more descriptive?

Comment (michelleangela) [reply Extension/src/LanguageServer/referencesProvider.ts]: also, should "isCandidates" be owned by the TreeNode object?

---
In reply to: [332144586](https://github.com/Microsoft/vscode-cpptools/pull/4378#discussion_r332144586) [](ancestors = 332144586)

File: Extension/src/LanguageServer/referencesProvider.ts [hunk 3]
```diff
@@ -47,119 +47,118 @@ function getReferenceItemIconPath(type: ReferenceType, isCanceled: boolean): { l
...
+                    return this.referencesModel.getRenameCandidateFiles();
+                } else {
+                    return this.referencesModel.getRenameCandidateReferenceTypes();
+                }
+            } else {
+                return this.referencesModel.getRenamePendingFiles();
+            }
         } else {
-            return this.referencesCanceled ? this.references.getReferenceCanceledGroup() : this.references.ReferenceTypeItems;
+            // Return root for tree view
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesProvider.ts]: >// Return root for tree view [](start = 12, length = 28)

update or remove comment

File: Extension/src/LanguageServer/referencesModel.ts [hunk 4]
```diff
@@ -6,126 +6,277 @@
...
-    readonly ReferenceTypeItems: ReferenceTypeItem[] = [];
+export type RenameResultCallback = (result: ReferencesResult) => void;
+
+let currentReferencesModel: ReferencesModel;
+
+export class ReferencesModel {
+    readonly nodes: TreeNode[] = []; // Raw flat list of references
+    private renameResultsCallback: RenameResultCallback;
     private originalSymbol: string = "";
+    public groupByFile: boolean;
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesModel.ts:17]: >groupByFile [](start = 11, length = 11)

can the view (referencesView) or referencesProvider own groupByFile property?

File: Extension/src/LanguageServer/referencesModel.ts [hunk 5]
```diff
@@ -6,126 +6,277 @@
...
+    // Optional properties for file related info
+    public filename?: string | undefined;
+    public fileUri?: vscode.Uri | undefined;
+
+    // Optional properties for reference item info
+    public referencePosition?: vscode.Position | undefined;
+    public referenceLocation?: vscode.Location | undefined;
+    public referenceText?: string | undefined;
+    public referenceType?: ReferenceType | undefined;
+    public isCandidate?: boolean | undefined;
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesModel.ts]: nit: change to isRenameCandidate

File: Extension/src/LanguageServer/referencesModel.ts [hunk 6]
```diff
@@ -6,126 +6,277 @@
...
+                    node.filename = n.filename;
+                    node.referenceType = n.referenceType;
+                    result.push(node);
+                }
+            }
+        });
+        return result;
+    }
+
+    cancelRename(): void {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesModel.ts]: should this be managed by references?

File: Extension/src/LanguageServer/referencesView.ts [hunk 7]
```diff
@@ -22,51 +23,59 @@ export class FindAllRefsView {
...
+                    confirmedRefs.push(line);
+                } else {
+                    otherRefs.push(line);
+                }
             }
         }
 
         // Get files with pending references items (location of reference is pending)
-        let fileReferences: FileItem[] = this.referenceViewProvider.getFilesWithPendingReferences();
+        let fileReferences: TreeNode[] = this.referencesModel.nodes.filter(i => (i.referencePosition.line === 0 && i.referencePosition.character === 0));
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesView.ts]: >this.referencesModel.nodes.filter(i => (i.referencePosition.line === 0 && i.referencePosition.character === 0)); [](start = 41, length = 112)

create function a getFileReferences() for consistency with other TreeNode item getters

File: Extension/src/LanguageServer/referencesView.ts [hunk 8]
```diff
@@ -6,13 +6,14 @@
...
-import { FileItem, ReferenceItem } from './referencesModel';
+import { ReferencesModel, NodeType, TreeNode } from './referencesModel';
 
 export class FindAllRefsView {
+    private referencesModel: ReferencesModel;
     private referenceViewProvider: ReferenceDataProvider;
 
     constructor() {
-        this.referenceViewProvider = new ReferenceDataProvider();
+        this.referenceViewProvider = new ReferenceDataProvider(false);
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesView.ts]: is this a View or Data provider? In rename it's called Data

File: Extension/src/LanguageServer/referencesView.ts [hunk 9]
```diff
@@ -22,51 +23,59 @@ export class FindAllRefsView {
...
-                ("[" + getReferenceTagString(ref.type, this.referenceViewProvider.isCanceled()) + "] "
-                + ref.parent.name
-                + ":" + (ref.position.line + 1) + ":" + (ref.position.character + 1)
-                + " " + ref.text);
-            if (includeConfirmedReferences && ref.type === ReferenceType.Confirmed) {
-                confirmedRefs.push(line);
-            } else {
-                otherRefs.push(line);
+        for (let ref of this.referencesModel.nodes) {
+            if (ref.node === NodeType.reference) {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesView.ts]: >if (ref.node === NodeType.reference) { [](start = 12, length = 38)

create a function that returns only NodeType.reference or use existing getReferenceNodes() function.

File: Extension/src/LanguageServer/referencesView.ts [hunk 10]
```diff
@@ -22,51 +23,59 @@ export class FindAllRefsView {
...
+                let line: string =
+                    ("[" + getReferenceTagString(ref.referenceType, this.referencesModel.isCanceled) + "] "
+                    + ref.filename
+                    + ":" + (ref.referencePosition.line + 1) + ":" + (ref.referencePosition.character + 1)
+                    + " " + ref.referenceText);
+                if (includeConfirmedReferences && ref.referenceType === ReferenceType.Confirmed) {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesView.ts]: >(includeConfirmedReferences && ref.referenceType === ReferenceType.Confirmed) [](start = 19, length = 77)

check "(includeConfirmedReferences && ref.referenceType === ReferenceType.Confirmed)" before creating a line string.

File: Extension/src/LanguageServer/referencesModel.ts [hunk 11]
```diff
@@ -6,126 +6,277 @@
...
-        group.push(refType);
-        return group;
+        return result;
     }
-}
 
-export class ReferenceTypeItem {
-    private files: FileItem[] = [];
+    getReferenceNodes(filename: string | undefined, refType: ReferenceType | undefined, isCandidate: boolean): TreeNode[] {
+        //let result: TreeNode[] = [];
```
Comment (michelleangela) [inline Extension/src/LanguageServer/referencesModel.ts]: >//let result: TreeNode[] = []; [](start = 8, length = 30)

remove commented out code

File: Extension/src/LanguageServer/referencesModel.ts (hunk 4 above)
Comment (Colengms) [reply Extension/src/LanguageServer/referencesModel.ts:17]: Some providers do not group by file (i.e. pending rename tree).  Also, isn't it preferable for the provider to be driven by data in the model?

File: Extension/src/LanguageServer/extension.ts [hunk 12]
```diff
@@ -1151,84 +1150,55 @@ function onRescanWorkspace(): void {
...
+    getCurrentRenameModel().completeRename();
 }
 
-function onRenameViewRemove(arg?: RenamePendingItem): void {
-    if (arg) {
-        arg.changeGroup();
-        arg.model.updateProviders();
-    }
+function onRenameViewRemove(arg?: TreeNode): void {
+    arg.model.setRenameCandidate(arg);
```
Comment (michelleangela) [inline Extension/src/LanguageServer/extension.ts:1181]: >arg.model.setRenameCandidate(arg); [](start = 4, length = 34)

check if arg defined?

File: Extension/src/LanguageServer/extension.ts [hunk 13]
```diff
@@ -1151,84 +1150,55 @@ function onRescanWorkspace(): void {
...
 }
 
-function onRenameViewAdd(arg?: RenameCandidateItem): void {
-    if (arg) {
-        arg.changeGroup();
-        arg.model.updateProviders();
-    }
+function onRenameViewAdd(arg?: TreeNode): void {
+    arg.model.setRenamePending(arg);
```
Comment (michelleangela) [inline Extension/src/LanguageServer/extension.ts:1188]: >arg.model.setRenamePending(arg); [](start = 4, length = 32)

check if arg defined? same for lines 1197 and 1201.

File: Extension/src/LanguageServer/extension.ts [hunk 14]
```diff
@@ -1151,84 +1150,55 @@ function onRescanWorkspace(): void {
...
-    }
+    getCurrentRenameModel().setAllRenamesPending();
 }
 
-function onRenameViewRemoveFile(arg?: RenamePendingFileItem): void {
-    if (arg) {
-        arg.changeGroup();
-        arg.model.updateProviders();
-    }
+function onRenameViewRemoveFile(arg?: TreeNode): void {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/extension.ts:1199]: >onRenameViewRemoveFile [](start = 9, length = 22)

onRenameViewRemoveFile() is empty?

**Pull request merged**